Physical_Engine/Query/IsWithinPileCap.cs

```diff
@@ -0,0 +1,112 @@
+using BH.Engine.Spatial;
+using BH.oM.Geometry;
+using BH.Engine.Geometry;
+
+namespace BH.Engine.Physical
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Determines whether the Piles are within the extents defined by the PadFoundation (both depth and extents). Piles located on the boundary are accepted within tolerance" +
+            "if their profile does not intersect with the pad foundation boundary.")]
+        [Input("padFoundation", "The PadFoundation that defines the extents for the Piles to be located.")]
+        [Input("piles", "The Piles to check if they are located within the extents of the PadFoundation.")]
+        [Input("tolerance", "Distance tolerance to be used in the method. Piles are deemed to be on the edge of the PadFoundation if they are within this distance from the curve.", typeof(Length))]
+        [Output("withinPileCap", "True if all Piles are contained within the boundary and depth of the PileCap. False if one or more Piles are not.")]
+        public static bool IsWithinPileCap(this PadFoundation padFoundation, List<Pile> piles, double tolerance = Tolerance.MicroDistance)
+        {
+            if (padFoundation.IsNull())
+                return false;
+            else if (piles == null)
+            {
+                Base.Compute.RecordError("The list of Piles are null, the method will return false.");
+                return false;
+            }
+            else if (piles.Any(x => x.IsNull()))
+                return false;
+
+            // Get the top of the piles
+            List<Point> tops = piles.Select(p => p.Location.IControlPoints().OrderBy(pt => pt.Z).Reverse().First()).ToList();
+
+            // Project all points to top surface of the pad
+            Plane padTop = new Plane() { Normal = padFoundation.Location.Normal(), Origin = padFoundation.Location.Centroid() };
+            List<Point> pTops = tops.Select(x => x.Project(padTop)).ToList(); //project points only used for IsContaining
+
+            // Get the thickness of the PadFoundation to compare against Z coordinates of the pile tops
+            double maxZ = padFoundation.Location.Centroid().Z;
+            double minZ = maxZ - padFoundation.Construction.IThickness();
+
+            ICurve topOutline = padFoundation.Location.ExternalBoundary;
+
+            // Check if all piles are within the curve of the pad
+            if(topOutline.IIsContaining(pTops,true,tolerance))
+            {
+                // Check if the piles are within the depth of the pad
+                if(tops.Any(pt => pt.Z > maxZ || pt.Z < minZ))
+                {
+                    Base.Compute.RecordError("One or more the Piles tops is not located within the depth of the PileCap.");
+                    return false;
+                }
```

Comment:
I made an attempt to enable tilted pads plus avoid computationally heavy containment check if possible - see below. Also worth checking if construction/thickness is set to the cap before running the check, and raise a warning about skipping the check if not?

```suggestion
            double padThickness = padFoundation.Construction.IThickness();
            for (int i = 0; i < tops.Count; i++)
            {
                if (tops[i].Distance(pTops[i]) + tolerance > padThickness)
                {
                    Base.Compute.RecordError("One or more the Piles tops is not located within the depth of the PileCap.");
                    return false;
                }
            }
            
            // check containment after making sure the piles are within the thickness
            if(topOutline.IIsContaining(pTops,true,tolerance))
```

Reply:
Have rearranged to include your suggestion - I don't think we can get away from including the `IsContaining()`.